E2e/update command for e2e

lanej0 opened the pull request:
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

Updates the command to run e2e tests to keep consistent with [the command for running API tests](https://github.com/woocommerce/woocommerce/pull/35283).

<!-- The next section is mandatory. If your PR doesn't require testing, please indicate that you are purposefully omitting instructions. -->

- [ ] This PR is a very minor change/addition and does not require testing instructions (if checked you can ignore/remove the next section).

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Otherwise, please include detailed instructions on how these changes can be tested (including pre-conditions, configuration, steps to take and expected results). It may help to write your instructions using pseudocode -- as if you're telling a computer how to execute the test. -->

1. Pull branch
2. `pnpm run env:test`
3. `pnpm test:e2e-pw`

Ensure that tests run normally (as if you had run `USE_WP_ENV=1 pnpx playwright test --config=tests/e2e-pw/playwright.config.js`)

<!-- End testing instructions -->

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm changelog add --filter=<project>`?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

Commit:
Update e2e test command for consistency

Commit:
README update

Commit:
Add changelog

zhongruige commented:
Thanks for getting this in @lanej0! It inspired me that I need to make some updates to the Playwright API Core Tests README as well. When I was going through it, I noticed a couple typos which are also on this README so wanted to share those here since we're updating it on this PR anyway:

* https://github.com/woocommerce/woocommerce/pull/35287/files#diff-7278e11f22f3f5e1233ee9eb52d3f04c038d89729ec989fd7ef38120cbe1374eR91
* https://github.com/woocommerce/woocommerce/pull/35287/files#diff-7278e11f22f3f5e1233ee9eb52d3f04c038d89729ec989fd7ef38120cbe1374eR95

zhongruige commented:
Unrelated to these changes, but we may need to consider how we'll need to handle releasing this as a package--we'll likely need to consider how we want to handle these instructions such as using NPM instead (e.g. `npm run test:e2e-pw`) or also providing the PNPM commands.

Came to mind as I was updating the [API tests readme](https://github.com/woocommerce/woocommerce/pull/35303/files) (thank you for the inspiration on this PR to do the same updates!) that eventually we'll want to consider how this looks outside of the monorepo context.

lanej0 committed:
Merge branch 'trunk' into e2e/update-command-for-e2e

tammullen commented on plugins/woocommerce/tests/e2e-pw/README.md:
```diff
@@ -31,57 +31,58 @@ End-to-end tests are powered by Playwright. The test site is spinned up using `w
-- `pnpm env:test --filter=woocommerce`
+-   `pnpm env:destroy`
+-   `pnpm env:test`
 
```
This needs additional line to actually do the re-run?
-   `pnpm test:e2e-pw`

tammullen commented on plugins/woocommerce/tests/e2e-pw/README.md:
```diff
@@ -31,57 +31,58 @@ End-to-end tests are powered by Playwright. The test site is spinned up using `w
+-   `pnpm test:e2e-pw --headed` (headed -- browser window shown)
+-   `pnpm test:e2e-pw --debug` (runs tests in debug mode)
+-   `pnpm test:e2e-pw ./tests/e2e-pw/tests/activate-and-setup/basic-setup.spec.js` (runs a single test)
 
```
Worth adding an example for debug for a single test as it's maybe more likely a use case for it?

tammullen commented on plugins/woocommerce/tests/e2e-pw/README.md:
```diff
@@ -93,50 +94,44 @@ Edit [.wp-env.json](https://github.com/woocommerce/woocommerce/blob/trunk/plugin
-- `pnpm env:destroy --filter=woocommerce` when you make changes to `.wp-env.json`
+-   `pnpm env:down --filter=woocommerce` to stop the environment
+-   `pnpm env:destroy --filter=woocommerce` when you make changes to `.wp-env.json`
 
```
Maybe include command to start environment?
`pnpm env:test` to spin up the test environment

alopezari replied:
What if we just add a `pnpm env:restart` that executed `env:destroy`and `env:test` inside?

Also, would it make sense to rename `env:test` to `env:start`? This would be aligned with the `wp-env` vocabulary, just as down and destroy do.

zhongruige replied:
Looping in @nigeljamesstevenson on this discussion as well, since we may want to consider updating the API tests documentation to align with these changes too.

nigeljamesstevenson replied:
Thanks for looping me on on this @zhongruige - i'll update the API tests documentation to align with these changes!

nigeljamesstevenson replied:
> Also, would it make sense to rename `env:test` to `env:start`? This would be aligned with the `wp-env` vocabulary, just as down and destroy do.

Cross referencing with the [package.json](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/package.json), if we rename `env:test` to `env:start` then it could be confused with `env:dev` which in turn calls `pnpm wp-env start`

we can also see that `pnpm env:down` in turn calls `pnpm wp-env stop` so it would be better to keep these consistent

Currently, the following could be updated:
`"env:dev": "pnpm wp-env start",` - could be changed to `"env:start": "pnpm wp-env start",`
`"env:test": "pnpm run env:dev && ./tests/e2e-pw/bin/test-env-setup.sh",` - could then be changed to `"env:test": "pnpm run env:start && ./tests/e2e-pw/bin/test-env-setup.sh",` (this would differentiate between the dev and test)
`"env:down": "pnpm wp-env stop",` - could be changed to `"env:stop": "pnpm wp-env stop", `

Whatever updates are made, I'll make sure the equivalent is updated for the `api-core-tests`
		
Thanks @tammullen @alopezari @zhongruige @lanej0 !

alopezari replied:
You're right @nigeljamesstevenson! I forgot about `pnpm env:dev`.

Maybe we can just do `pnpm env:dev:start` to start the dev environment and `pnpm env:test:start` to start the test one. And also apply the related changes to stop and destroy.

I just suggested this because if you read `pnpm env:test` out of context, you might think that this command "tests" something, but it just starts the test environment.

This might make the structure of the commands a bit more complex, so I'd like to know your thoughts here too @nigeljamesstevenson @tammullen @zhongruige @lanej0, in case I'm missing something and complicating things 😄

zhongruige replied:
Perhaps we could go with something simple like this?

`pnpm env:up`

alopezari replied:
We would be in a similar situation, isn't it? We should differentiate between dev and test, so it would look like
`pnpm env:dev:up` and `pnpm env:test:up`

Ps: both up and start look fine to me 👍 the "problem" is considering if we would be OK to add a new layer to the command name (moving from env:test to env:test:start or env:test:up).

zhongruige replied:
Hmm I feel like chaining that way makes it a bit more complex, I am not sure if we need the `dev` in there. So if we just had the following:

* `pnpm env:start`
* `pnpm env:stop`
* `pnpm env:destroy`

Perhaps to avoid the `env:test` issue, maybe we just rename this to something like `pnpm env:setup` (or configure or something like that) since it just [sets up the environment](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/tests/e2e-pw/bin/test-env-setup.sh) required for the tests. This way everything environment related is under `env`.

With the above, then we have our commands for the tests ran against the env and for any new tests, since it's a bit more semantically clear that these run tests and nothing specifically with the env:

* `pnpm test:e2e-pw`
* `pnpm test:api-pw`
* `pnpm test:k6`

alopezari replied:
I like this!

If we don't need to keep the dev environment in mind for this naming convention, then we can avoid that extra layer of complexity in the command, so as you said, this would look like:
```
pnpm env:start
pnpm env:stop
pnpm env:destroy
```
Or even replace `pnpm env:start` with `pnpm env:setup`, as you suggested, both look fine to me 👍 

Thanks @zhongruige!

nigeljamesstevenson replied:
I agree - the less complexity the better!

```
pnpm env:start
pnpm env:stop
pnpm env:destroy
```
which all tie in nicely with the `wp-env` commands that they are calling in the package.json and then also have
`pnpm env:setup`
for the setting up of the test environment i.e. this starts the environment with `pnpm env:start` and then calls the `test-env-setup.sh` for the additional test env setup

i.e. I would keep `pnpm env:start` and not replace it 😊

We would just have to make sure that any updates have previous references to them updated. e.g. `pnpm env:dev` is used in [workflows](https://github.com/woocommerce/woocommerce/blob/7cb84e5b2d6c16da30ef876271854e59e3581912/plugins/woocommerce/tests/performance/README.md#:~:text=tests%20in%20action.-,pnpm%20run%20env%3Adev,-%2D%2Dfilter%3Dwoocommerce%0Apnpm) and [readmes](https://github.com/woocommerce/woocommerce/blob/7cb84e5b2d6c16da30ef876271854e59e3581912/plugins/woocommerce/tests/performance/README.md#:~:text=tests%20in%20action.-,pnpm%20run%20env%3Adev,-%2D%2Dfilter%3Dwoocommerce%0Apnpm) in addition to the [package.json](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/package.json#:~:text=%22-,env%3Adev,-%22%3A%20%22)

Thanks @alopezari  and @zhongruige 👏🏻

tammullen commented:
> I noticed a couple typos which are also on this README so wanted to share those here since we're updating it on this PR anyway:

Link to PR as the other link didn't work for me https://github.com/woocommerce/woocommerce/pull/35303/files
Agree it would be good to address these, although not a blocker if we don't want to do it within this PR

alopezari commented on plugins/woocommerce/tests/e2e-pw/README.md:
```diff
@@ -93,50 +94,44 @@ Edit [.wp-env.json](https://github.com/woocommerce/woocommerce/blob/trunk/plugin
 
 **Modiify port for e2e-environment**
```
There is a typo in `Modiify`, I know it wasn't introduced in this change but considering there are pending actions after @tammullen comments, I think it would be a great opportunity to fix this if you don't mind @lanej0 🙏

Commit:
Merge branch 'trunk' into e2e/update-command-for-e2e

# Conflicts:
#	plugins/woocommerce/tests/e2e-pw/README.md

Commit:
Incorporated feedback from review

lanej0 commented:
I *think* I synthesized the feedback from this review into a set of sensible changes.  To summarize:

- Updated README with proper commands/instuctions
- Updated package.json and added:

`pnpm env:start`
`pnpm env:stop`
`pnpm env:restart` (which does a env:destroy && env:start)

`pnpm env:destroy` already exists

I didn't remove the old env:down env:dev or any of the old puppeteer scripts.  Things are getting a little messy in there, so we should follow up with a PR in the future to clean some of that out.

lanej0 commented:
@alopezari @tammullen @nigeljamesstevenson -- tagging you all as you participated in this discussion (and apologies in advance for letting this PR languish).  If I've missed anything, let me know!

Commit:
Need to run config after starting

alopezari commented:
Thanks, @lanej0!

Looks good to me 👍 

Adding @woocommerce/atlas to the review as monorepo owners for a final check, just in case we missed anything.

Also, the new `env:start` and `env:restart` commands make use of `wp-env start` and there is another command just for this purpose `env:dev`. Should we create a follow-up task in GitHub to review and clean up outdated or duplicated commands?

nigeljamesstevenson commented:
Great stuff @lanej0 !

In order to avoid some duplication of commands:
`"env:restart": "pnpm wp-env destroy && pnpm wp-env start && ./tests/e2e-pw/bin/test-env-setup.sh",`

could be changed to:
`"env:restart": "pnpm wp-env destroy && pnpm env:start",`

nigeljamesstevenson commented:
This looks good to me. I have created a task to consolidate the commands [here](https://github.com/woocommerce/woocommerce/issues/36833)

samueljseay reviewed (approved):
Looks good from my perspective 🚀
